src/lib.rs

```diff
@@ -958,36 +933,22 @@ impl<'a> WebViewBuilder<'a> {
   ///   folder which lives within the apk. For the cases where this can be used, it works the same as in macOS and Linux.
   /// - iOS: To get the path of your assets, you can call [`CFBundle::resources_path`](https://docs.rs/core-foundation/latest/core_foundation/bundle/struct.CFBundle.html#method.resources_path). So url like `wry://assets/index.html` could get the html file in assets directory.
   #[cfg(feature = "protocol")]
-  pub fn with_custom_protocol<F>(self, name: String, handler: F) -> Self
+  pub fn with_custom_protocol<F>(mut self, name: String, handler: F) -> Self
   where
     F: Fn(WebViewId, Request<Vec<u8>>) -> Response<Cow<'static, [u8]>> + 'static,
   {
-    self.and_then(|mut b| {
-      #[cfg(any(
-        target_os = "linux",
-        target_os = "dragonfly",
-        target_os = "freebsd",
-        target_os = "netbsd",
-        target_os = "openbsd",
-      ))]
-      if let Some(context) = &mut b.attrs.context {
-        context.register_custom_protocol(name.clone())?;
-      }
-
-      if b.attrs.custom_protocols.iter().any(|(n, _)| n == &name) {
-        return Err(Error::DuplicateCustomProtocol(name));
-      }
-
-      b.attrs.custom_protocols.insert(
-        name,
-        Box::new(move |id, request, responder| {
-          let http_response = handler(id, request);
-          responder.respond(http_response);
-        }),
-      );
-
-      Ok(b)
-    })
+    if self.attrs.custom_protocols.contains_key(&name) {
+      self.error = Err(crate::Error::DuplicateCustomProtocol(name));
```

Comment:
in the future there maybe other methods added that would require returning an err, so I think instead of replacing `self.error` we should chain `self.error.and_then`